src/network/network_chat_gui.cpp

```diff
@@ -346,95 +344,85 @@ struct NetworkChatWindow : public Window {
 			for (const Town *t : Town::Iterate(*item - MAX_CLIENT_SLOTS)) {
 				/* Get the town-name via the string-system */
 				SetDParam(0, t->index);
-				GetString(chat_tab_temp_buffer, STR_TOWN_NAME, lastof(chat_tab_temp_buffer));
-				return &chat_tab_temp_buffer[0];
+				return GetString(STR_TOWN_NAME);
 			}
 		}
 
-		return nullptr;
+		return std::nullopt;
 	}
 
 	/**
 	 * Find what text to complete. It scans for a space from the left and marks
 	 *  the word right from that as to complete. It also writes a \0 at the
 	 *  position of the space (if any). If nothing found, buf is returned.
 	 */
-	static char *ChatTabCompletionFindText(char *buf)
+	static std::string_view ChatTabCompletionFindText(std::string_view &buf)
 	{
-		char *p = strrchr(buf, ' ');
-		if (p == nullptr) return buf;
+		auto it = buf.find_last_of(' ');
+		if (it == std::string_view::npos) return buf;
 
-		*p = '\0';
-		return p + 1;
+		std::string_view res = buf.substr(it + 1);
+		buf.remove_suffix(res.size() + 1);
+		return res;
 	}
 
 	/**
 	 * See if we can auto-complete the current text of the user.
 	 */
 	void ChatTabCompletion()
 	{
-		static char _chat_tab_completion_buf[NETWORK_CHAT_LENGTH];
-		assert(this->message_editbox.text.max_bytes == lengthof(_chat_tab_completion_buf));
+		static std::string _chat_tab_completion_buf;
 
 		Textbuf *tb = &this->message_editbox.text;
-		size_t len, tb_len;
-		uint item;
-		char *tb_buf, *pre_buf;
-		const char *cur_name;
+		uint item = 0;
 		bool second_scan = false;
 
-		item = 0;
-
 		/* Copy the buffer so we can modify it without damaging the real data */
-		pre_buf = (_chat_tab_completion_active) ? stredup(_chat_tab_completion_buf) : stredup(tb->buf);
-
-		tb_buf  = ChatTabCompletionFindText(pre_buf);
-		tb_len  = strlen(tb_buf);
+		std::string_view pre_buf = _chat_tab_completion_active ? std::string_view(_chat_tab_completion_buf) : std::string_view(tb->buf);
+		std::string_view tb_buf = ChatTabCompletionFindText(pre_buf);
 
-		while ((cur_name = ChatTabCompletionNextItem(&item)) != nullptr) {
+		std::optional<std::string> cur_item;
+		while ((cur_item = ChatTabCompletionNextItem(&item)).has_value()) {
+			std::string_view cur_name = cur_item.value();
 			item++;
 
 			if (_chat_tab_completion_active) {
 				/* We are pressing TAB again on the same name, is there another name
 				 *  that starts with this? */
 				if (!second_scan) {
-					size_t offset;
-					size_t length;
+					std::string_view view;
 
 					/* If we are completing at the begin of the line, skip the ': ' we added */
-					if (tb_buf == pre_buf) {
-						offset = 0;
-						length = (tb->bytes - 1) - 2;
+					if (tb_buf.data() == pre_buf.data()) {
```

Comment:
Yes, comparing the pointers is needed. Otherwise something like 'Hi: Hi:<tab complete>' is going to trigger this branch, as both `tb_buf` and `pre_buf` have the same content... though they're not pointing to the same location in the original string.

Reply:
One could argue that this would be a perfectly good code comment 😄